assumeutxo: Check snapshot base block is not in invalid chain

fjahr opened the pull request:
This was discovered in a discussion in #29996

If the base block of the snapshot is marked invalid or part of an invalid chain, we currently still load the snapshot and get stuck in a weird state where we have the snapshot chainstate but it will never connect to our valid chain.

While this scenario is highly unlikely to occur on mainnet, it still seems good to prevent this inconsistent state.

The behavior change described above is in the second commit.

The first commit refactors the early checks in the `loadtxoutset` RPC by moving them into `ActivateSnapshot()` in order to have the chance to cover them by unit tests in the future and have a more consistent interface. Previously checks were spread out between `rpc/blockchain.cpp` and `validation.cpp`. In order to be able to return the error message to users of the RPC, the return type of `ActivateSnapshot()` is changed from `bool` to `util::Result`.

The third commit removes an unnecessary restart introduced in #29428.

tdb3 commented on test/functional/feature_assumeutxo.py:
```diff
@@ -202,7 +202,15 @@ def test_snapshot_with_less_work(self, dump_output_path):
+        self.log.info("Test snapshot is not loaded when base block is invalid.")
+        node = self.nodes[0]
+        block_hash = node.getblockhash(SNAPSHOT_BASE_HEIGHT - 1)
+        node.invalidateblock(block_hash)
```
> If the base block of the snapshot is marked invalid or part of an invalid chain

This existing test seems to focus on the base block being part of an invalid chain.  Is it also worth adding an explicit test for an invalid base block itself (or would it be overkill)?

fjahr replied:
I think it's a bit overkill but it was easy to add and also shouldn't make the test considerably slower, so I added coverage for this as well.

tdb3 reviewed (approved):
ACK cbc604f6a7739bf22a5d811b6366453860d59f4c
Seems like an improvement to avoid expending resources loading a snapshot that won't be part of a valid chain.

Built and ran unit and functional tests (including `feature_assumeutxo`).  All passed.

Tested that `feature_assumeutxo` would fail with:
```c++
bool start_block_invalid = false;
```
Test failed as expected.

alfonsoromanz reviewed (approved):
tACK cbc604f6a7739bf22a5d811b6366453860d59f4c

This PR successfully builds and passes the test when running `test/functional/feature_assumeutxo.py`.

Additionally, I tested this in the context of my PR ([#29996](https://github.com/bitcoin/bitcoin/pull/29996)) back when the code looked something like this:

```
    def test_snapshot_in_a_divergent_chain(self, dump_output_path):
        node = self.nodes[2]
        # Rollback the chain down to START_HEIGHT
        block_hash = node.getblockhash(START_HEIGHT + 1)
        node.invalidateblock(block_hash)
        assert_equal(node.getblockcount(), START_HEIGHT)

        self.log.info(f"Check importing a snapshot where current chain-tip is not an ancestor of the snapshot block but has less work")
        # Generate a divergent chain up to 298
        self.generate(node, nblocks=99, sync_fun=self.no_op)
        assert_equal(node.getblockcount(), SNAPSHOT_BASE_HEIGHT - 1)

        # Try importing the snapshot and assert its success
        self.log.info('Importing the snapshot')
        loaded = node.loadtxoutset(dump_output_path)
```

The snapshot load failed with the expected message: `test_framework.authproxy.JSONRPCException: The base block header is part of an invalid chain. (-32603)`

BrandonOdiwuor reviewed (commented):
Code Review ACK cbc604f6a7739bf22a5d811b6366453860d59f4c

BrandonOdiwuor commented on src/rpc/blockchain.cpp:
```diff
@@ -2836,6 +2836,15 @@ static RPCHelpMan loadtxoutset()
+    if (start_block_invalid) {
+        throw JSONRPCError(
+            RPC_INTERNAL_ERROR,
+            "The base block header is part of an invalid chain.");
```
Nit: Would it be better to include the base block hash on the error message to help in debugging 

```cpp
throw JSONRPCError(
            RPC_INTERNAL_ERROR,
            "The base block header (%s) is part of an invalid chain.", base_blockhash.ToString());
```

fjahr replied:
taken with minor edits

mzumsande commented on src/rpc/blockchain.cpp:
```diff
@@ -2836,6 +2836,15 @@ static RPCHelpMan loadtxoutset()
+
+    bool start_block_invalid = WITH_LOCK(::cs_main,
+            return snapshot_start_block->nStatus & BLOCK_FAILED_MASK);
+    if (start_block_invalid) {
```
Some checks are within `ActivateSnapshot()`, others, like the added one, are in the rpc code (which means that they can't be covered in  unit tests for that function). Do you know if there is a reason for that?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30267#discussion_r1638706909

> Some checks are within `ActivateSnapshot()`, others, like the added one, are in the rpc code (which means that they can't be covered in unit tests for that function). Do you know if there is a reason for that?

I think it this is just a mistake, and it makes the checks less reliable because they are done without keeping cs_main locked. I suggested fixing this before in https://github.com/bitcoin/bitcoin/pull/27596#discussion_r1325193433.

fjahr replied:
I think these were always separated like that. But sounds good to me to move them. I have added a refactoring commit which moves the other two early checks into `ActivateSnapshot()` and then I am adding the new check there as well.

mzumsande replied:
Looks good (except for the compiler issue with `Join()`) - I think it would be helpful to not only log the reason, but also return them with the RPC - for that, `ActivateSnapshot()` would need to return it, possibly as `util::Result` as suggested by ryanofsky, but that could be a follow-up.
Though for now, it might be helpful to expand the error ("Unable to load UTXO snapshot") to point the user to the debug log for the reason.

fjahr replied:
Ah, yeah, I gave it a shot and it pretty simple so I am changing the return type as well in the refactoring commit now.

fjahr committed:
test: Remove unnecessary restart in assumeutxo test

fjahr committed:
refactor: Move early loadtxoutset checks into ActiveSnapshot

Also changes the return type of ActiveSnapshot to allow returning the
error message to the user of the loadtxoutset RPC.

fjahr committed:
assumeutxo: Check snapshot base block is not marked invalid

Co-authored-by: Alfonso Roman Zubeldia <alfonsoromanz24@gmail.com>

fjahr commented:
Needed to rebase because of a silent merge conflict.

mzumsande commented on src/rpc/blockchain.cpp:
```diff
@@ -2821,34 +2819,15 @@ static RPCHelpMan loadtxoutset()
-        throw JSONRPCError(RPC_INTERNAL_ERROR, "Unable to load UTXO snapshot " + fs::PathToString(path));
+    auto activation_result{chainman.ActivateSnapshot(afile, metadata, false)};
+    if (!activation_result) {
+        throw JSONRPCError(RPC_INTERNAL_ERROR, strprintf(_("Unable to load UTXO snapshot: %s\n"), util::ErrorString(activation_result)).original);
```
This changes it such that the path is no longer included in the RPC error. I think that it wasn't really helpful anyway (the user already provided the path when calling the RPC, so why the need to report it back to the user?), but just wanted to mention it because in cases someone would miss that.

mzumsande reviewed (commented):
Code Review ACK 750f8b50a749e577fc11f2f9f79e06cdd29e66f5

fjahr commented:
Just fixed typo noticed by @mzumsande : https://github.com/bitcoin/bitcoin/compare/750f8b50a749e577fc11f2f9f79e06cdd29e66f5..2f9bde69f45c7a9fdcf0c65f9e1305391a6f1f28

mzumsande commented:
re-ACK 2f9bde6

alfonsoromanz commented:
Re-ACK 2f9bde69f45c7a9fdcf0c65f9e1305391a6f1f28. The RPC code looks much cleaner after the refactor. Also, it seems very useful to get the error message in the RPC response rather than having to rely on the logs in some scenarios if you are an RPC user.

achow101 commented:
ACK 2f9bde69f45c7a9fdcf0c65f9e1305391a6f1f28

maflcko reviewed (commented):
post-merge lgtm. Left some nits

ryanofsky commented on src/validation.cpp:
```diff
@@ -5729,7 +5743,7 @@ bool ChainstateManager::ActivateSnapshot(
             }
         }
-        return false;
+        return util::Error{_(reason)};
```
re: https://github.com/bitcoin/bitcoin/pull/30267#discussion_r1663631973

> It is not possible to pass a raw c-string pointer to transifex for translation. Only string literals can be translated.

In case it's not clear, the fix for this problem and the previous [unnecessary translation](https://github.com/bitcoin/bitcoin/pull/30267#discussion_r1663611194) problem is probably to replace `_()` with `Untranslated()` everywhere where in this PR when converting strings to bilingual_str. If we added a snapshot loading GUI in the future, it might make sense to add translations at that point.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30267#discussion_r1663631973

> It is not possible to pass a raw c-string pointer to transifex for translation. Only string literals can be translated.

Fixed in #30417

alfonsoromanz commented:
Post-merge question: Shouldn't this TODO be deleted because of the new test introduced in this PR?

> - TODO: Valid snapshot file, but referencing a snapshot block that turns out to be
>       invalid, or has an invalid parent

The `test_snapshot_block_invalidated` function seems to address this scenario by invalidating the snapshot base block and one of its parents, ensuring the snapshot cannot be loaded if the base block or its parent is invalid.

fjahr commented:
> Post-merge question: Shouldn't this TODO be deleted because of the new test introduced in this PR?

Duh, you are right, fixed with #30403. Thanks!

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -2821,34 +2819,15 @@ static RPCHelpMan loadtxoutset()
-        throw JSONRPCError(RPC_INTERNAL_ERROR, "Unable to load UTXO snapshot " + fs::PathToString(path));
+    auto activation_result{chainman.ActivateSnapshot(afile, metadata, false)};
+    if (!activation_result) {
+        throw JSONRPCError(RPC_INTERNAL_ERROR, strprintf(_("Unable to load UTXO snapshot: %s\n"), util::ErrorString(activation_result)).original);
```
re: https://github.com/bitcoin/bitcoin/pull/30267#discussion_r1663611194

> What is the point of translating this message when the translation is discarded? Seems wasteful of translator's time, no?

Removed translation in #30417

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -2821,34 +2819,15 @@ static RPCHelpMan loadtxoutset()
-        throw JSONRPCError(RPC_INTERNAL_ERROR, "Unable to load UTXO snapshot " + fs::PathToString(path));
+    auto activation_result{chainman.ActivateSnapshot(afile, metadata, false)};
+    if (!activation_result) {
+        throw JSONRPCError(RPC_INTERNAL_ERROR, strprintf(_("Unable to load UTXO snapshot: %s\n"), util::ErrorString(activation_result)).original);
```
re: https://github.com/bitcoin/bitcoin/pull/30267#discussion_r1663647981

> Is the trailing newline needed? This would be the first `JSONRPCError` with a trailing newline.

Removed in #30417
